File: PresenterCorner/wwwroot/js/presentationhub.js
const connection = new signalR.HubConnectionBuilder()
    .withUrl("/presentationHub")
    .build()
    .connection.start().then(() => {
    console.log("Connected to SignalR Hub");
    });

let currentPresentationId = null;
let currentUserRole = "Viewer";
let nickname = prompt("Enter your nickname:");

connection.on("ReceiveUpdate", (update) => {
    // Handle real-time updates here
    console.log("Update received:", update);
    applyUpdate(update);
});

connection.on("UserJoined", (user) => {
    console.log("User joined:", user);
    addUserToList(user);
});

connection.on("RoleAssigned", (newRole) => {
    currentUserRole = newRole;
    alert(`Your role has been updated to: ${newRole}`);
});

// Triggered when user joins a presentation
function joinPresentation(presentationId, nickname) {
    currentPresentationId = presentationId;
    await connection.invoke("JoinPresentation", presentationId, nickname);
}

function applyUpdate(update) {
    // Example: Apply a text block update
    const { type, data } = update;
    if (type === "text-block") {
        const { id, content, position } = data;
        let textBlock = document.getElementById(id);
        if (!textBlock) {
            textBlock = document.createElement("div");
            textBlock.id = id;
            textBlock.className = "text-block";
            document.getElementById("activeSlide").appendChild(textBlock);
        }
        textBlock.textContent = content;
        textBlock.style.top = `${position.top}px`;
        textBlock.style.left = `${position.left}px`;
    }
}

function addUserToList(user) {
    const userList = document.getElementById("userList");
    const listItem = document.createElement("li");
    listItem.className = "list-group-item";
    listItem.textContent = `${user.nickname} (${user.role})`;
    userList.appendChild(listItem);
}

document.getElementById("addSlideBtn").addEventListener("click", () => {
    if (currentUserRole !== "Creator") {
        alert("Only the creator can add slides.");
        return;
    }

    const slideId = `slide-${Date.now()}`;
    const update = { type: "add-slide", slideId };
    connection.invoke("SendUpdate", currentPresentationId, update);
});

// Start the connection
connection.start().then(() => {
    console.log("Connected to SignalR hub.");
    const presentationId = prompt("Enter presentation ID to join:");
    joinPresentation(presentationId);
});

// Triggered when an editor modifies an element on a slide
function sendSlideUpdate(presentationId, updateData, userRole) {
    if (userRole === "Viewer") {
        console.error("Viewers cannot edit slides.");
        return;
    }

    connection.invoke("SendUpdate", presentationId, updateData, userRole)
        .then(() => {
            console.log("Update sent:", updateData);
        })
        .catch(err => console.error(err.toString()));
}

// Triggered when the creator assigns a role
//function assignRole(connectionId, role) {
//    connection.invoke("AssignRole", connectionId, role)
//        .then(() => {
//            console.log("Role assigned:", connectionId, role);
//        })
//        .catch(err => console.error(err.toString()));
//}

File: PresenterCorner/Hubs/PresentationHub.cs
using Microsoft.AspNetCore.SignalR;
using Microsoft.EntityFrameworkCore;
using PresenterCorner.Data;
using PresenterCorner.Models;
using System;

namespace PresenterCorner.Hubs
{
   public class PresentationHub : Hub
   {
        private readonly ApplicationDbContext _context;

        public PresentationHub(ApplicationDbContext context)
        {
            _context = context;
        }

        public override async Task OnConnectedAsync()
        {
            var connectionId = Context.ConnectionId;
            var userId = Context.GetHttpContext().Request.Query["userId"];

            var user = await _context.Users.FindAsync(int.Parse(userId));
            if (user != null)
            {
                user.ConnectionId = connectionId;
                await _context.SaveChangesAsync();
            }
            // Optionally handle user connection logic
            await base.OnConnectedAsync();
        }
        // Notify all clients in a presentation about an update
        public async Task SendUpdate(string presentationId, object update, string role)
        {
            if (role == "Viewer")
            {
                throw new HubException("Viewers cannot send updates.");
            }
            await Clients.Group(presentationId).SendAsync("ReceiveUpdate", update);

        }
        // Add a user to a presentation group
        public async Task JoinPresentation(string presentationId, string nickname)
        {
            var user = _context.Users.FirstOrDefault(u => u.ConnectionId == Context.ConnectionId);
            if (user == null)
            {
                user = new User
                {
                    ConnectionId = Context.ConnectionId,
                    Nickname = nickname,
                    Role = "Viewer" // Default role
                };
                _context.Users.Add(user);
                await _context.SaveChangesAsync();
            }

            await Groups.AddToGroupAsync(Context.ConnectionId, presentationId);
            await Clients.Group(presentationId).SendAsync("UserJoined", user);

        }
        // Remove a user from a presentation group
        public async Task LeavePresentation(string presentationId)
        {
            await Groups.RemoveFromGroupAsync(Context.ConnectionId, presentationId);
        }

        public async Task AssignRole(string connectionId, string role)
        {
            var user = _context.Users.FirstOrDefault(u => u.ConnectionId == connectionId);
            if (user != null)
            {
                user.Role = role;
                await _context.SaveChangesAsync();
            }
            await Clients.Client(connectionId).SendAsync("RoleAssigned", role);
        }
    }

}
